components/user-mgt/org.wso2.carbon.identity.user.pre.update.profile.action/src/main/java/org/wso2/carbon/identity/user/pre/update/profile/action/internal/execution/PreUpdateProfileRequestBuilder.java

```diff
@@ -0,0 +1,327 @@
+    }
+
+    private User getUser(UserActionContext userActionContext, PreUpdateProfileAction preUpdateProfileAction,
+                         UniqueIDUserStoreManager userStoreManager) throws ActionExecutionRequestBuilderException {
+
+        UserActionRequestDTO userActionRequestDTO = userActionContext.getUserActionRequestDTO();
+        List<String> userClaimsToSetInEvent = preUpdateProfileAction.getAttributes();
+
+        User.Builder userBuilder = new User.Builder(userActionRequestDTO.getUserId());
+        if (userClaimsToSetInEvent == null || userClaimsToSetInEvent.isEmpty()) {
+            return userBuilder.build();
+        }
+
+        try {
+            Map<String, String> claimValues = userStoreManager.getUserClaimValuesWithID(
+                    userActionRequestDTO.getUserId(), userClaimsToSetInEvent.toArray(new String[0]),
+                    UserCoreConstants.DEFAULT_PROFILE);
+
+            String multiAttributeSeparator = FrameworkUtils.getMultiAttributeSeparator();
+            setClaimsInUserBuilder(userBuilder, claimValues, userActionRequestDTO.getClaims(), multiAttributeSeparator);
+            setGroupsInUserBuilder(userBuilder, claimValues, multiAttributeSeparator);
+        } catch (org.wso2.carbon.user.core.UserStoreException e) {
+            throw new ActionExecutionRequestBuilderException("Failed to retrieve user claims from user store.", e);
+        }
+        return userBuilder.build();
+    }
+
+    private void setClaimsInUserBuilder(User.Builder userBuilder, Map<String, String> claimValues,
+                                        Map<String, Object> updatingUserClaimsInRequest,
+                                        String multiAttributeSeparator) throws ActionExecutionRequestBuilderException {
+
+        List<UpdatingUserClaim> userClaimValuesToSetInEvent = new ArrayList<>();
+        for (Map.Entry<String, String> claimEntry : claimValues.entrySet()) {
+            String claimKey = claimEntry.getKey();
+            String claimValue = claimEntry.getValue();
+
+            if (isRoleOrGroupClaim(claimKey) || StringUtils.isBlank(claimValue)) {
+                continue;
+            }
+
+            if (isMultiValuedClaim(claimKey)) {
+                userClaimValuesToSetInEvent.add(
+                        constructMultiValuedClaim(updatingUserClaimsInRequest, claimKey, claimValue,
+                                multiAttributeSeparator));
+            } else {
+                userClaimValuesToSetInEvent.add(constructSingleValuedClaim(updatingUserClaimsInRequest, claimKey,
+                        claimValue));
+            }
+        }
+
+        userBuilder.claims(userClaimValuesToSetInEvent);
+    }
+
+    private void setGroupsInUserBuilder(User.Builder userBuilder, Map<String, String> claimValues,
+                                        String multiAttributeSeparator) {
+
+        if (claimValues.get(GROUP_CLAIM_URI) != null) {
```

Comment:
Do we have groups in this groups claim URL. We handle groups claim differently compared to other attributes. Hence let's double check about this.

Reply:
groups are available in groups claim URI. 
And this groups claim URI is how groups gets communicated to the application.
Is there any place this is handled differently ?